src/string/grapheme/grapheme.cr

```diff
@@ -0,0 +1,258 @@
+  # thumbs up sign `👍` (`U+1F44D`) combined with an emoji modifier such as
+  # `U+1F3FC` assign a colour to the emoji.
+  #
+  # Instances of this type can be acquired via `String#each_grapheme` or `String#graphemes`.
+  #
+  # The algorithm to determine boundaries between grapheme clusters is specified
+  # in the [Unicode Standard Annex #29](https://www.unicode.org/reports/tr29/tr29-37.html#Grapheme_Cluster_Boundaries),
+  # and implemented in Version Unicode 13.0.0.
+  struct Grapheme
+    # For efficency reasons we avoid allocating a string for graphemes consisting
+    # of only a single character.
+    # As a trade-off, this leads to multi dispatch on this ivar. But that's
+    # acceptable compared to the allocation overhead.
+    #
+    # Graphemes consisting of a single character are always represented as
+    # `Char`, never as `String` to simplify comparability. This invariant is
+    # protected by the constructor.
+    @cluster : Char | String
+
+    # :nodoc:
+    def self.new(string : String, range : Range(Int32, Int32), char : Char) : self
+      if char.bytesize == range.size
+        new(char)
+      else
+        new(string.byte_slice(range.begin, range.end - range.begin))
+      end
+    end
+
+    # :nodoc:
+    def initialize(@cluster)
+    end
+
+    # Appends the characters in this grapheme cluster to *io*.
+    def to_s(io : IO) : Nil
+      io << @cluster
+    end
+
+    # Returns the characters in this grapheme cluster.
+    def to_s : String
+      case cluster = @cluster
+      in Char
+        cluster.to_s
+      in String
+        cluster
+      end
+    end
+
+    # Appends a representation of this grapheme cluster to *io*.
+    def inspect(io : IO) : Nil
+      io << "String::Grapheme("
+      @cluster.inspect(io)
+      io << ")"
+    end
+
+    # Returns the number of characters in this grapheme cluster.
+    def size : Int32
+      case cluster = @cluster
+      in Char
+        1
+      in String
+        cluster.size
+      end
+    end
+
+    # Returns the number of bytes in the UTF-8 representation of this grapheme cluster.
+    def bytesize : Int32
+      @cluster.bytesize
+    end
+
+    # Returns `true` if *other* is equivalent to `self`.
+    #
+    # Two graphemes are considered equivalent if they contain the same sequence
+    # of codepoints.
+    def ==(other : self) : Bool
+      @cluster == other.@cluster
+    end
+
+    # :nodoc:
+    def self.break?(c1 : Char, c2 : Char) : Bool
+      break?(Property.from(c1), Property.from(c2))
+    end
+
+    # :nodoc:
+    #
+    # Returns whether there is a grapheme break between boundclasses lbc and tbc.
+    #
+    # Please note that evaluation of GB10 (grapheme breaks between emoji zwj sequences)
+    # and GB 12/13 (regional indicator code points) require knowledge of previous characters
+    # which is not handled by this oberload. This may result in an incorrect break before
```

Comment:
```suggestion
    # which is not handled by this overload. This may result in an incorrect break before
```

Reply:
I read it as _uberlord_ in the first sec 😅